### PR TITLE
Define rename-symbol contract and expand capability framework

### PR DESCRIPTION
**Pull request by leynos**
## Summary
Introduce a formal capability contract system for plugin capabilities (starting with rename-symbol) and expand the plugin ecosystem to support capability declarations, contract validation, and capability-based discovery. This includes a new capability module, typed request validation, stable reason codes, and registry lookups, along with tests and documentation updates.

## Changes
- Add capability contract framework
  - New module: crates/weaver-plugins/src/capability/mod.rs
    - CapabilityId enum (RenameSymbol, ExtricateSymbol, ExtractMethod, ReplaceBody, ExtractPredicate)
    - ContractVersion struct with major/minor and compatibility checks
    - CapabilityContract trait for request/response validation
    - Re-exports for consumer crates
  - New module: crates/weaver-plugins/src/capability/reason_code.rs
    - ReasonCode enum with stable snake_case strings and serde support
  - New module: crates/weaver-plugins/src/capability/rename_symbol.rs
    - RENAME_SYMBOL_CONTRACT_VERSION constant (1.0)
    - RenameSymbolRequest with typed fields and extraction/validation from PluginRequest
    - RenameSymbolContract implementing CapabilityContract with validation rules
    - Helpers for validating successful diff output in responses
  - New tests: crates/weaver-plugins/src/capability/tests.rs
    - Serde round-trips, contract validation, version compatibility, reason codes
- Extend core protocol/types
  - crates/weaver-plugins/src/protocol/mod.rs
    - Add optional reason_code: ReasonCode to PluginDiagnostic with serde defaults
    - Builder-style with_reason_code(code) and accessor reason_code()
- Extend manifest handling
  - crates/weaver-plugins/src/manifest/mod.rs
    - Add capabilities: Vec<CapabilityId> field with serde(default)
    - with_capabilities(...) builder, capabilities() accessor
    - Sensor vs actuator guard: sensors must not declare actuator capabilities
- Extend registry lookups
  - crates/weaver-plugins/src/registry/mod.rs
    - find_for_capability(id: CapabilityId) -> Vec<&PluginManifest>
    - find_for_language_and_capability(language: &str, id: CapabilityId) -> Vec<&PluginManifest>
- Wiring and public API surface
  - crates/weaver-plugins/src/lib.rs
    - pub mod capability;
  - Re-export capability types for easier downstream use
- Tests and behaviour-driven tests
  - Unit tests for capability contract and naming in crates/weaver-plugins/src/capability/tests.rs
  - Capability-based lookup tests in crates/weaver-plugins/src/registry/tests.rs
  - Behaviour tests scaffold: crates/weaver-plugins/src/tests/capability_behaviour.rs
  - BDD feature: crates/weaver-plugins/tests/features/capability_contract.feature
  - BDD step definitions: crates/weaver-plugins/src/tests/capability_behaviour.rs
- Documentation and docs updates
  - docs/execplans/5-2-1-define-the-rename-symbol-capability-contract.md (new)
  - docs/users-guide.md: section detailing plugin capabilities, rename-symbol contract, reason codes, and manifest example
  - docs/roadmap.md: mark 5.2.1 as completed
- Execution plan and examples
  - New capability-focused capability_contract feature scenarios to exercise contract validation, refusals, and manifest capabilities

## Rationale
- Provides a robust, versioned contract system for capabilities (starting with rename-symbol) to enable safe broker-plugin interaction.
- Introduces stable, programmatic diagnostics via ReasonCode to support automation and tooling.
- Enables capability-based routing and discovery via manifest capabilities and registry lookups.
- Ensures backwards compatibility by using serde(default) for new fields and by validating only supported shapes in contracts.

## Compatibility and Migration
- Backwards compatible: new fields are optional/defaulted so existing manifests and diagnostics deserialize without changes.
- Manifest validation enforces that sensor plugins do not declare actuator capabilities, preserving existing semantics.
- No breaking changes to existing public API beyond the addition of new capabilities and fields; existing code should continue to function.

## Testing plan
- Unit tests covering: capability IDs, contract versioning, request/response validation, and reason-code serialization.
- Registry tests for capability-based lookups and language capability intersections.
- Behaviour-driven tests (BDD) for end-to-end contract validation, refusals with reason codes, and manifest capability handling.
- All tests should pass with cargo test; lint/check-fmt should remain green.

## Documentation and users' guide updates
- Document the capability model, the five first-party capabilities, and the rename-symbol contract in the users guide.
- Provide examples of manifest capability declarations and how reason codes appear in diagnostics.
- Update roadmap to reflect completion of 5.2.1 and related notes.

## Review tips
- Focus on the contract implementation, especially validate_request and validate_response paths for rename-symbol.
- Verify compatibility rules in ContractVersion and how is_compatible_with is used by the broker.
- Check that manifest/sensor validation logic correctly rejects sensor plugins with actuator capabilities.
- Validate new registry lookups behave as expected for capability-based discovery and language filtering.

If you want me to tweak the PR message (tone, depth, or section order), tell me your preferences and I’ll adjust accordingly.

📎 **Task**: https://www.devboxer.com/task/6720b407-5feb-4589-8ee8-2a3572c2ae4a

## Summary by Sourcery

Introduce a versioned capability contract system (starting with rename-symbol) for actuator plugins, including manifest capability declarations, capability-based registry lookup, structured refusal reason codes on diagnostics, and corresponding tests and documentation updates.

New Features:
- Add a capability module with identifiers, contract versioning, a rename-symbol capability contract, and stable refusal reason codes.
- Allow plugin manifests to declare supported capabilities and expose them via builder and accessor methods.
- Extend the plugin registry with capability and language-aware lookup functions.
- Support optional reason_code fields on plugin diagnostics for programmatic error handling.
- Add behaviour-driven tests and feature scenarios for capability contracts and validation flows.

Enhancements:
- Document plugin capabilities, the rename-symbol contract, and reason codes in the user guide and execution plan docs, and mark the related roadmap item as complete.

Tests:
- Add unit tests for capability identifiers, contract versioning, rename-symbol request/response validation, reason codes, manifest validation, registry capability lookups, and diagnostic reason_code behaviour.
- Add BDD feature files and step definitions exercising capability contract validation, refusal handling, manifest capability rules, and version compatibility.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-plugins/src/manifest/tests.rs](https://github.com/leynos/weaver/pull/67/files/0ed14e84ff6a3776d6877df8c881b205fbc82ecf#diff-351aaeeb3fa7a075d5dfc7830a23ef3ba704f01b0f8db6bc9c87f5c9f2f1b8f0)

Comment on file

~~~rust
fn manifest_without_capabilities_deserialises_to_empty() {
    let json = r#"{
        "name": "test",
        "version": "0.1",
        "kind": "actuator",
        "languages": ["rust"],
        "executable": "/bin/test"
    }"#;
    let m: PluginManifest = serde_json::from_str(json).expect("deserialise");
    assert!(m.capabilities().is_empty());
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=67&biomarker=Code+Duplication&filename=crates%2Fweaver-plugins%2Fsrc%2Fmanifest%2Ftests.rs&method=)
The module contains 2 functions with similar structure: manifest_deserialise_defaults_timeout,manifest_without_capabilities_deserialises_to_empty

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-plugins/src/manifest/tests.rs](https://github.com/leynos/weaver/pull/67/files/a7a2cb14b6de8fa37bc6c03c5261c84e58db7b02#diff-351aaeeb3fa7a075d5dfc7830a23ef3ba704f01b0f8db6bc9c87f5c9f2f1b8f0)

Comment on file

~~~rust
    assert_manifest_from_json(json, |m| {
        assert_eq!(m.timeout_secs(), 30);
        assert!(m.args().is_empty());
    });
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=67&biomarker=Code+Duplication&filename=crates%2Fweaver-plugins%2Fsrc%2Fmanifest%2Ftests.rs&method=)
The module contains 2 functions with similar structure: manifest_deserialise_defaults_timeout,manifest_without_capabilities_deserialises_to_empty

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- In `RenameSymbolContract::validate_request`, you currently validate only the arguments map and ignore `PluginRequest::operation`; consider asserting that the operation matches the capability (e.g. `"rename-symbol"`) so that obvious routing or wiring bugs are caught early rather than assuming the caller passed the right operation string.
- The registry helpers `find_for_capability` and `find_for_language_and_capability` scan all manifests and repeatedly allocate lowercase language strings on each call; if you expect many plugins or frequent lookups, consider normalising languages and/or indexing manifests by `(language, CapabilityId)` at registration time to avoid repeated per-call work.

## Individual Comments

### Comment 1
<location path="crates/weaver-plugins/src/manifest/mod.rs" line_range="203-208" />
<code_context>
                 ),
             });
         }
+        if self.kind == PluginKind::Sensor && !self.capabilities.is_empty() {
+            return Err(PluginError::Manifest {
+                message: String::from("sensor plugins must not declare actuator capabilities"),
+            });
+        }
</code_context>
<issue_to_address>
**suggestion:** Validation error message is more specific than the actual check being performed.

The current guard bans all capabilities on `Sensor` plugins, while the error text only mentions actuator capabilities. Please either generalize the message to match the check, or narrow the check to explicitly reject only actuator‑only capabilities (if that distinction exists in the type system).

```suggestion
        }
        if self.kind == PluginKind::Sensor && !self.capabilities.is_empty() {
            return Err(PluginError::Manifest {
                message: String::from("sensor plugins must not declare any capabilities"),
            });
        }
```
</issue_to_address>

### Comment 2
<location path="crates/weaver-plugins/src/capability/mod.rs" line_range="154-168" />
<code_context>
+    /// Returns the contract version.
+    fn version(&self) -> ContractVersion;
+
+    /// Validates that a plugin request conforms to the contract schema.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`PluginError::InvalidOutput`] if the request does not
+    /// conform to the expected schema.
+    fn validate_request(&self, request: &PluginRequest) -> Result<(), PluginError>;
+
+    /// Validates that a plugin response conforms to the contract schema.
</code_context>
<issue_to_address>
**suggestion:** The documented error type for request/response validation may be misleading.

The docs currently promise `PluginError::InvalidOutput` for request validation failures, but a request/schema issue is conceptually an input error. If there is (or might be) a more appropriate `PluginError` variant, consider using that instead. Otherwise, you could phrase this as "returns a `PluginError` if the request/response does not conform" without naming a specific variant.

```suggestion
    /// Validates that a plugin request conforms to the contract schema.
    ///
    /// # Errors
    ///
    /// Returns a [`PluginError`] if the request does not
    /// conform to the expected schema.
    fn validate_request(&self, request: &PluginRequest) -> Result<(), PluginError>;

    /// Validates that a plugin response conforms to the contract schema.
    ///
    /// # Errors
    ///
    /// Returns a [`PluginError`] if a successful response
    /// does not conform to the expected output schema.
    fn validate_response(&self, response: &PluginResponse) -> Result<(), PluginError>;
```
</issue_to_address>

### Comment 3
<location path="crates/weaver-plugins/src/tests/capability_behaviour.rs" line_range="71" />
<code_context>
+/// Each token is split on the first `=`; the value portion has surrounding
+/// double-quotes stripped. This is intentionally simple because the inputs
+/// are controlled test fixtures.
+fn parse_kv_pairs(input: &str) -> HashMap<String, serde_json::Value> {
+    let mut map = HashMap::new();
+    for token in input.split_whitespace() {
</code_context>
<issue_to_address>
**suggestion (testing):** Add unit tests for `parse_kv_pairs` to avoid silently misparsing BDD step arguments

This helper is a critical dependency for BDD step parsing but currently has no direct tests. To prevent regressions (e.g., around empty input, duplicate keys, or values with `=`) and to surface parsing issues clearly, please add targeted unit tests in this module that cover: (1) empty string → empty map, (2) multiple key/value pairs, and (3) tokens without `=` being ignored.
</issue_to_address>

### Comment 4
<location path="docs/execplans/5-2-1-define-the-rename-symbol-capability-contract.md" line_range="155" />
<code_context>
+## Decision log
+
+- Decision: Use an enum for `CapabilityId` rather than a string-backed
+  newtype. Rationale: ADR 001 defines exactly five first-party capability IDs
+  (`rename-symbol`, `extricate-symbol`, `extract-method`, `replace-body`,
+  `extract-predicate`). An enum provides compile-time exhaustiveness checking,
</code_context>
<issue_to_address>
**issue (review_instructions):** ADR is introduced without expansion; expand it (e.g. “Architecture Decision Record (ADR) 001”) on first use.

ADR is an uncommon acronym in this context. To follow the style guidance, spell it out on first use, for example: “Architecture Decision Record (ADR) 001 defines exactly five first-party capability IDs …”. Subsequent references can use just “ADR 001”.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 5
<location path="docs/execplans/5-2-1-define-the-rename-symbol-capability-contract.md" line_range="31" />
<code_context>
+- New unit tests in `crates/weaver-plugins/src/capability/` exercise schema
+  validation for happy and unhappy paths.
+- New BDD scenarios in
+  `crates/weaver-plugins/tests/features/capability_contract.feature` exercise
+  contract validation, refusal diagnostics, and manifest capability
+  declarations.
</code_context>
<issue_to_address>
**issue (review_instructions):** BDD is used without being defined; expand it on first occurrence (e.g. “behaviour-driven development (BDD)”).

BDD is not expanded anywhere before this line in the document. Please define it on first use, for example: “New behaviour-driven development (BDD) scenarios in …”. Later mentions can use the acronym alone.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: crates/weaver-plugins/src/capability/rename_symbol.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    /// # Errors
+    ///
+    /// Returns [`PluginError::InvalidOutput`] if required fields are
+    /// missing or have invalid types.
+    pub fn extract(request: &PluginRequest) -> Result<Self, PluginError> {
+        let args = request.arguments();
+
+        let uri = extract_string_field(args, "uri")?;
+        let position = extract_string_field(args, "position")?;
+        let new_name = extract_string_field(args, "new_name")?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/capability/rename_symbol.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaver-plugins/src/capability/rename_symbol.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    fn capability_id(&self) -> CapabilityId {
+        CapabilityId::RenameSymbol
+    }
+
+    fn version(&self) -> ContractVersion {
+        RENAME_SYMBOL_CONTRACT_VERSION
+    }
+
+    fn validate_request(&self, request: &PluginRequest) -> Result<(), PluginError> {
+        RenameSymbolRequest::extract(request).map(|_| ())
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/capability/rename_symbol.rs:180]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaver-plugins/src/capability/tests.rs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+fn contract_validate_wrong_operation_rejects(contract: RenameSymbolContract) {
+    let args = make_rename_args("file:///src/main.py", "10:5", "bar");
+    let request = PluginRequest::with_arguments("extricate-symbol", vec![], args);
+    let err = contract
+        .validate_request(&request)
+        .expect_err("should reject wrong operation");
+    assert!(
+        err.to_string().contains("expects operation"),
+        "expected operation mismatch error, got: {err}",
+    );
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/src/capability/tests.rs:257]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

File: crates/weaver-plugins/tests/features/capability_contract.feature [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  Scenario: Contract version compatibility with same major
+    Given contract version 1.0
+    And a peer version 1.3
+    Then the versions are compatible
+
+  Scenario: Contract version incompatibility with different major
+    Given contract version 1.0
+    And a peer version 2.0
+    Then the versions are incompatible
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugins/tests/features/capability_contract.feature:75]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

File: docs/execplans/5-2-1-define-the-rename-symbol-capability-contract.md [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+   doc comment for the rename-symbol contract. `RenameSymbolRequest` struct
+   defining the typed request shape with `extract()` method that validates the
+   untyped arguments HashMap. `RenameSymbolContract` struct implementing
+   `CapabilityContract`. `RENAME_SYMBOL_CONTRACT_VERSION` constant. Validation
+   logic for request and response.
+
+3. `crates/weaver-plugins/src/capability/reason_code.rs` — Module-level doc
+   comment for refusal reason codes. `ReasonCode` enum with seven stable codes
+   serialized as snake_case strings.
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/execplans/5-2-1-define-the-rename-symbol-capability-contract.md:313]: @coderabbitai "serialized" is correct for en-GB-oxendict (-ize / -yse / -our)
